mining/src/mempool/model/transactions_pool.rs

```diff
@@ -181,70 +187,49 @@ impl TransactionsPool {
-    /// to have room for at least `free_slots` new transactions. The returned transactions
+    /// to make room for `transaction`. The returned transactions
     /// are guaranteed to be unchained (no successor in mempool) and to not be parent of
     /// `transaction`.
     ///
-    /// An error is returned if the mempool is filled with high priority transactions.
-    pub(crate) fn limit_transaction_count(
-        &self,
-        free_slots: usize,
-        transaction: &MutableTransaction,
-    ) -> RuleResult<Vec<TransactionId>> {
-        assert!(free_slots > 0);
+    /// An error is returned if the mempool is filled with high priority transactions, or
+    /// at least one candidate for removal has a higher fee rate than `transaction`.
+    pub(crate) fn limit_transaction_count(&self, transaction: &MutableTransaction) -> RuleResult<Vec<TransactionId>> {
         // Returns a vector of transactions to be removed that the caller has to remove actually.
         // The caller is golang validateAndInsertTransaction equivalent.
         // This behavior differs from golang impl.
-        let trim_size = self.len() + free_slots - usize::min(self.len() + free_slots, self.config.maximum_transaction_count as usize);
-        let mut transactions_to_remove = Vec::with_capacity(trim_size);
-        if trim_size > 0 {
-            // TODO: consider introducing an index on all_transactions low-priority items instead.
-            //
-            // Sorting this vector here may be sub-optimal compared with maintaining a sorted
-            // index of all_transactions low-priority items if the proportion of low-priority txs
-            // in all_transactions is important.
-            let low_priority_txs = self
-                .all_transactions
-                .values()
-                .filter(|x| x.priority == Priority::Low && self.transaction_is_unchained(&x.id()) && !x.is_parent_of(transaction));
-
-            if trim_size == 1 {
-                // This is the most likely case. Here we just search the minimum, thus avoiding the need to sort altogether.
-                if let Some(tx) = low_priority_txs.min_by(|a, b| a.fee_rate().partial_cmp(&b.fee_rate()).unwrap()) {
-                    transactions_to_remove.push(tx);
-                }
-            } else {
-                let mut low_priority_txs = low_priority_txs.collect::<Vec<_>>();
-                if low_priority_txs.len() > trim_size {
-                    low_priority_txs.sort_by(|a, b| a.fee_rate().partial_cmp(&b.fee_rate()).unwrap());
-                    transactions_to_remove.extend_from_slice(&low_priority_txs[0..usize::min(trim_size, low_priority_txs.len())]);
-                } else {
-                    transactions_to_remove = low_priority_txs;
-                }
+        let transaction_size = transaction.tx.estimate_mem_bytes();
+        let mut txs_to_remove = Vec::with_capacity(1);
```

Comment:
Why `Vec::with_capacity(1)`? Add a comment as to why. My guess is that the expected number of transactions to remove is almost always exactly `1`.